src/languages/go.js

```diff
@@ -109,10 +109,29 @@ export default function(hljs) {
         className: 'number',
         variants: [
           {
-            begin: hljs.C_NUMBER_RE + '[i]',
-            relevance: 1
+            begin: '-?\\b0[xX]\\.[a-fA-F0-9](_?[a-fA-F0-9])*[pP][+-]?\\d(_?\\d)*i?', // hex without a present digit before . (making a digit afterwards required)
+            relevance: 0
           },
-          hljs.C_NUMBER_MODE
+          {
+            begin: '-?\\b0[xX](_?[a-fA-F0-9])+((\\.([a-fA-F0-9](_?[a-fA-F0-9])*)?)?[pP][+-]?\\d(_?\\d)*)?i?', // hex with a present digit before . (making a digit afterwards optional)
```

Comment:
Please use the newer `match` (instead of `begin`)... and please rewrite these to be literal regex, rather than escaped strings.

Also any reason we can't group the two octals together?